File: com/example/truthmoji/truthmoji.zig
// TruthMoji implementation in Zig
const std = @import("std");

// TruthMoji struct with atomically incremented usage counter
pub const TruthMoji = struct {
    emoji_true: []const u8,
    emoji_false: []const u8,
    usage_count: std.atomic.Value(usize),

    // Constructor
    pub fn init() TruthMoji {
        return TruthMoji{
            .emoji_true = "😊",
            .emoji_false = "😔",
            .usage_count = std.atomic.Value(usize).init(0),
        };
    }

    // Convert boolean to emoji
    pub fn emojify(self: *TruthMoji, value: bool) []const u8 {
        // Thread-safe increment
        _ = self.usage_count.fetchAdd(1, .monotonic);
        
        return if (value) self.emoji_true else self.emoji_false;
    }

    // Get usage count
    pub fn getUsage(self: *const TruthMoji) usize {
        return self.usage_count.load(.monotonic);
    }
};

// Global instance for module-level access
var g_instance = TruthMoji.init();

// Module API functions
pub fn emojify(value: bool) []const u8 {
    return g_instance.emojify(value);
}

pub fn getUsage() usize {
    return g_instance.getUsage();
}

pub fn main() !void {
    // Demo functionality
    const stdout = std.io.getStdOut().writer();
    
    try stdout.print("=== TruthMoji Demo (Zig) ===\n", .{});
    try stdout.print("true → {s}\n", .{emojify(true)});
    try stdout.print("false → {s}\n", .{emojify(false)});
    try stdout.print("1==1 → {s}\n", .{emojify(1 == 1)});
    try stdout.print("1==0 → {s}\n", .{emojify(1 == 0)});
    try stdout.print("Usage: {d}\n", .{getUsage()});
    
    // Print instance details
    try stdout.print("Inspect: TruthMoji {{ emoji_true: {s}, emoji_false: {s}, usage_count: {d} }}\n", 
        .{g_instance.emoji_true, g_instance.emoji_false, g_instance.getUsage()});
}

File: com/example/truthmoji/truthmoji.rs
// TruthMoji implementation in Rust
use std::sync::atomic::{AtomicUsize, Ordering};

// TruthMoji struct with atomic counter for thread safety
struct TruthMoji {
    emoji_true: String,
    emoji_false: String,
    usage_count: AtomicUsize,
}

impl TruthMoji {
    // Constructor
    fn new() -> Self {
        TruthMoji {
            emoji_true: "😊".to_string(),
            emoji_false: "😔".to_string(),
            usage_count: AtomicUsize::new(0),
        }
    }
    
    // Convert boolean to emoji
    fn emojify(&self, value: bool) -> &str {
        // Thread-safe increment
        self.usage_count.fetch_add(1, Ordering::SeqCst);
        
        if value {
            &self.emoji_true
        } else {
            &self.emoji_false
        }
    }
    
    // Get usage count
    fn get_usage(&self) -> usize {
        self.usage_count.load(Ordering::SeqCst)
    }
    
    // Inspect internal state
    fn inspect(&self) -> String {
        format!(
            "TruthMoji {{ emoji_true: {}, emoji_false: {}, usage_count: {} }}",
            self.emoji_true, self.emoji_false, self.get_usage()
        )
    }
}

fn main() {
    // Create a new TruthMoji instance
    let tm = TruthMoji::new();
    
    // Demo functionality
    println!("=== TruthMoji Demo (Rust) ===");
    println!("true → {}", tm.emojify(true));
    println!("false → {}", tm.emojify(false));
    println!("1==1 → {}", tm.emojify(1 == 1));
    println!("1==0 → {}", tm.emojify(1 == 0));
    println!("Usage: {}", tm.get_usage());
    println!("Inspect: {}", tm.inspect());
}